### PR TITLE
Community Health: monthly community health report system

### LINKED ISSUE #4329: Community health: alert system for PMs
As part of our recent push to unify tools for PMs and centralize/improve locale health detection, we are looking to implement a monthly notification/email report based on `LocaleHealthSnapshot` deltas.

This will give PMs and admins a "mini dashboard" type of overview for certain locales and hopefully prevent locale decline through early detection.

### PR DESCRIPTION
**Pull request by functionzz**
This PR creates a community health alert system for PMs, and allows PMs see month-over-month changes to CHS at a glance. It is compiled monthly.

image

Fixes #4329.

### CODE REVIEW

File: pontoon/contributors/templates/contributors/widgets/notifications_menu.html [hunk 1]
```diff
@@ -87,7 +87,7 @@
           {% endif %}
 
           {% if description and (description.startswith("Unreviewed suggestions") or notification.verb == "has reviewed suggestions" or notification.verb == "ignore") %}
-            <span class="description">{{ description|safe }}</span>
+            <div class="description">{{ description|safe }}</div>
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/widgets/notifications_menu.html]: What's the purpose of this change?

File: pontoon/base/static/css/style.css
Comment (mathjazz) [inline pontoon/base/static/css/style.css:1]: What's the purpose of these changes?

File: pontoon/base/migrations/0130_userprofile_monthly_health_report.py
Comment (mathjazz) [inline pontoon/base/migrations/0130_userprofile_monthly_health_report.py:1]: We already have migration 0125 (and 0126).

File: pontoon/messaging/tests/test_emails.py [hunk 2]
```diff
@@ -215,6 +217,46 @@ def test_send_inactive_translator_emails(user_a, locale_a):
     assert mail.outbox[0].to == [user_a.contact_email]
 
 
+@pytest.mark.django_db
```
Comment (mathjazz) [inline pontoon/messaging/tests/test_emails.py:220]: We should also add a test for the case when none of the locales passes a threshold.

**Review by mathjazz — COMMENTED** [review]
Nice work!

Some high-level questions:
- Why is the /insights page shown to and the notification sent to `is_staff` users , but the emails can be opted in by `is_superuser` users? 
- Why don't we automatically opt in all `is_staff` users also for receiving emails and drop the settings? This is an altert system afterall. Or at least opt them in by default, if we *really* want to keep the ability to opt out.
- Have we tested the notification in the translate view, which has its own notification implementation?
- The docstring says rows are "ranked by the size of that change", but the biggest gains come first and the worst regressions land at the bottom of the table. In an alert system, the regressions are the actionable half. Sort by `abs(percentage)` to match the docstring?
- The new command needs to be documented on `documentation/docs/dev/deployment.md`. We should mention it silently depends on cron ordering - it reads the current month's snapshot — which `collect_chs_snapshots` creates on that same day.
- We should also document the reporting feature for the Admins. Note that there's still https://github.com/mozilla/pontoon/pull/4338#issuecomment-5163970110.

**Comment by flodolo** [comment]
> * Why don't we automatically opt in all `is_staff` users also for receiving emails and drop the settings? This is an altert system afterall. Or at least opt them in by default, if we _really_ want to keep the ability to opt out.

I don't think all superusers are PMs interested in this information, so they should have a way to opt out. I have the feeling the majority falls into this category, hence the opt-in suggestion to Jamie.

File: pontoon/contributors/templates/contributors/widgets/notifications_menu.html (hunk 1 above)
Comment (functionzz) [reply pontoon/contributors/templates/contributors/widgets/notifications_menu.html]: If we keep it as a span, if the `description` contains html elements, they will all be converted to text, which is not what we want.

File: pontoon/base/static/css/style.css
Comment (functionzz) [reply pontoon/base/static/css/style.css:1]: Because I changed description to a div, I also want it to inherit css styling as well.

File: pontoon/contributors/templates/contributors/widgets/notifications_menu.html (hunk 1 above)
Comment (functionzz) [reply pontoon/contributors/templates/contributors/widgets/notifications_menu.html]: The main point is now we have a system notification that automatically sends but also contains more complex html + css, which was not supported with the current code. All our previous system notifications were just one-liner spans.

File: pontoon/base/migrations/0130_userprofile_monthly_health_report.py
Comment (functionzz) [reply pontoon/base/migrations/0130_userprofile_monthly_health_report.py:1]: ~do I need to rebase? I don't understand the comment.~ I see

**Comment by functionzz** [comment]
> Why is the /insights page shown to and the notification sent to is_staff users , but the emails can be opted in by is_superuser users?

I actually think it should be only reserved to is_staff users - I made the change before you reviewed, but it seems I didn't completely change options. ~I need to change both user pool to be from is_staff.~ Change complete.


> Why don't we automatically opt in all is_staff users also for receiving emails and drop the settings? This is an alert system after all. Or at least opt them in by default, if we really want to keep the ability to opt out.

Just like the other email systems in Pontoon, I think we should keep the option to opt in given both the notification & email output very similar things.

> Have we tested the notification in the translate view, which has its own notification implementation?

~Not yet, will do that now.~ Merged css + `<span>` -> `<div>`changes similar to django notifications.

> The docstring says rows are "ranked by the size of that change", but the biggest gains come first and the worst regressions land at the bottom of the table. In an alert system, the regressions are the actionable half. Sort by abs(percentage) to match the docstring?

Sure. Implemented.

> The new command needs to be documented on documentation/docs/dev/deployment.md. We should mention it silently depends on cron ordering - it reads the current month's snapshot — which collect_chs_snapshots creates on that same day. We should also document the reporting feature for the Admins. Note that there's still https://github.com/mozilla/pontoon/pull/4338#issuecomment-5163970110.

Sure, I'll make the changes in this PR. Implemented.

File: pontoon/messaging/tests/test_emails.py (hunk 2 above)
Comment (functionzz) [reply pontoon/messaging/tests/test_emails.py:220]: Done.

**Comment by functionzz** [comment]
I've been testing out the 2% threshold for CHS, and I'm beginning to think that is not sufficient, as some Common Voice locales may have deltas greater than 2% but these changes an insignificantly small. Perhaps changing to a 2 point threshold makes more sense?

File: pontoon/contributors/templates/contributors/settings.html [hunk 3]
```diff
@@ -110,6 +110,18 @@ <h3>Email</h3>
             title='Receive monthly activity summary email',
             help="Get an email summary of your personal activity and the activity within your teams in the last month.")
           }}
+
+          {% if user.is_superuser %}
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/settings.html]: Looks like this one slipped through?

```suggestion
          {% if user.is_staff %}
```

**Comment by mathjazz** [comment]
> I don't think all superusers are PMs interested in this information, so they should have a way to opt out. I have the feeling the majority falls into this category, hence the opt-in suggestion to Jamie.

OK. We should at least file an issue then to introduce a "PM" user role.

**Review by mathjazz — COMMENTED** [review]
Thanks for the update!

Could you please unify the copy between the notification and the email and update the screenshot with the latest version?

**Comment by mathjazz** [comment]
> I've been testing out the 2% threshold for CHS, and I'm beginning to think that is not sufficient, as some Common Voice locales may have deltas greater than 2% but these changes an insignificantly small. Perhaps changing to a 2 point threshold makes more sense?

How did we get to 2% in the first place? Have we looked at the old data?

**Comment by functionzz** [comment]
> How did we get to 2% in the first place?

We didn't really discuss it, this was something I more or less added as a placeholder.

File: pontoon/messaging/templates/messaging/emails/monthly_health_report.html [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+{% block content %}
+  <p>
+    The monthly locale health report for {{ month }} {{ year }} is now
+    available.
+  </p>
+
+  {% if locale_rows %}
+    <p>
+      Below are the locales whose Community Health Score (CHS) changed by at
+      least {{ threshold }}% compared to the previous month.
```
Comment (mathjazz) [inline pontoon/messaging/templates/messaging/emails/monthly_health_report.html]:
```suggestion
      Locales whose Community Health Score (CHS) changed by at least
      {{ threshold }}% compared to the previous month and have at least
      one key project enabled:
```

File: pontoon/messaging/templates/messaging/emails/monthly_health_report.html [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+              {{ locale_row.percentage }}
+            </td>
+          </tr>
+        {% endfor %}
+      </tbody>
+    </table>
+
+    <p class="note">
+      Only locales with at least one key project enabled are included.
+    </p>
```
Comment (mathjazz) [inline pontoon/messaging/templates/messaging/emails/monthly_health_report.html]:
```suggestion
```

Moved that to the paragraph above.

File: pontoon/messaging/templates/messaging/emails/monthly_health_report.html [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+{% endblock %}
+
+{% block content_class %}monthly-health-report{% endblock %}
+
+{% block content %}
+  <p>
+    The monthly locale health report for {{ month }} {{ year }} is now
+    available.
+  </p>
+
```
Comment (mathjazz) [inline pontoon/messaging/templates/messaging/emails/monthly_health_report.html]:
```suggestion
```

Can we drop this? It doesn't add much value compared to the title.

File: pontoon/messaging/templates/messaging/emails/monthly_health_report.html [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    </table>
+
+    <p class="note">
+      Only locales with at least one key project enabled are included.
+    </p>
+  {% else %}
+    <p>
+      However, no locale's Community Health Score (CHS) changed by at least
+      {{ threshold }}% compared to the previous month.
+    </p>
```
Comment (mathjazz) [inline pontoon/messaging/templates/messaging/emails/monthly_health_report.html:88]:
```suggestion
  {% else %}
    <p>
      There were no locales whose Community Health Score (CHS) changed by at least
      {{ threshold }}% compared to the previous month.
    </p>
```

**Review by mathjazz — COMMENTED** [review]
Left some notes to make the copy a bit more condensed.

File: documentation/docs/admin/insights.md [hunk 8]
```diff
@@ -1,35 +1,57 @@
 # Insights
 
+Pontoon provides 2 major systems for admins to monitor locale & project health. The tools
+available are the `Insights` dashboard and the `Monthly locale health report`, which both provide
+information on locale growth & decline in easily ingestible formats.
+
+## Insights dashboard
+
```
Comment (mathjazz) [inline documentation/docs/admin/insights.md:8]: I would drop this section, which would allow us to keep the existing heading levels. 4 levels seems like a lot for a relatively short page.

File: documentation/docs/admin/insights.md [hunk 9]
```diff
@@ -1,35 +1,57 @@
...
 Plots the approval rate of pretranslations for each team, making it possible to compare pretranslation quality across locales and spot teams whose custom machine translation models may need attention.
 
-### Project pretranslation quality
+#### Project pretranslation quality
 
 Plots the approval rate of pretranslations for each project, highlighting which projects produce the most and least reliable pretranslations.
+
+## Monthly locale health report
+
+Once a month, after new locale health snapshots are collected, Pontoon compares the two most recent `Community health scores` of each team and reports the ones that moved the most. The report lists every affected team with its score from the previous month, its current score, and the change between them, expressed as a percentage of the previous score. Teams are ordered by the size of that change, largest first, regardless of whether they went up or down.
```
Comment (mathjazz) [inline documentation/docs/admin/insights.md]:
```suggestion
Once a month, after new locale health snapshots are collected, Pontoon compares each team’s two most recent Community Health Scores and reports the teams with the largest changes. The report lists each affected team’s score from the previous month, its current score, and the percentage change relative to the previous score. Teams are ordered by the absolute size of the change, largest first, regardless of whether their scores increased or decreased.
```

File: documentation/docs/admin/insights.md [hunk 10]
```diff
@@ -1,35 +1,57 @@
...
+
+A team is included when it is enabled for at least one [key project](#key-projects), has a snapshot in both months, and its score changed by at least 2%. The threshold can be changed through the `MONTHLY_HEALTH_REPORT_CHS_THRESHOLD` environment variable.
```
Comment (mathjazz) [inline documentation/docs/admin/insights.md]:
```suggestion
A team is included if it is enabled for at least one [key project](#key-projects), has a snapshot for both months, and its score changed by at least 2%. The threshold can be configured using the `MONTHLY_HEALTH_REPORT_CHS_THRESHOLD` environment variable.
```

File: documentation/docs/admin/insights.md [hunk 11]
```diff
@@ -1,35 +1,57 @@
...
+
+Staff users receive the report as a notification in Pontoon. To also receive it by email, enable `Monthly locale health report` in the `Email` section of your [settings](../localizer/users.md).
```
Comment (mathjazz) [inline documentation/docs/admin/insights.md]: We link to the Settings page, but the new setting is not documented there.

File: documentation/docs/admin/insights.md [hunk 12]
```diff
@@ -1,35 +1,57 @@
...
 
-### Show scores
+#### Show scores
 
 By default the table shows the raw value of each metric. Click the `Show scores` button to switch to the score view, which instead displays the individual component scores that add up to the `Community health score`. Click the button again (now labeled `Show default`) to return to the raw values.
 
-## Community health score chart
+#### Key projects
+
+Two of the metrics above — number of enabled projects and project completion — are not measured across everything a team works on, but only across the projects marked as **key projects**. Enabled projects is scored out of a share of total projects, while project completion is the share of translated strings within the key projects the team is enabled for. Disabled projects never count toward either metric, and project completion ignores system projects and projects that are not visible.
```
Comment (mathjazz) [inline documentation/docs/admin/insights.md]:
```suggestion
Two of the metrics above — the number of enabled projects and project completion — are calculated only across projects marked as key projects, rather than across all projects a team works on. The enabled projects metric is the share of all key projects enabled for the team. The project completion metric is the share of translated strings across the key projects for which the team is enabled. Disabled projects do not count toward either metric, while project completion also excludes system projects and projects that are not visible.
```

File: documentation/docs/admin/insights.md [hunk 13]
```diff
@@ -1,35 +1,57 @@
...
+
+Marking a project as a key project is available in the Django admin (`/a/base/project/`) using the `Is chs project` flag. A team then counts as enabled for that project as soon as the team is added to it, the same way as for any other project (see [Adding a new project](adding-new-project.md)).
```
Comment (mathjazz) [inline documentation/docs/admin/insights.md]:
```suggestion
Administrators can mark a project as a key project in the Django admin (`/a/base/project/`) using the `Is chs project` flag. A team counts as enabled for a key project as soon as it is added to that project, just as it would for any other project (see [Adding a new project](adding-new-project.md)).
```

File: documentation/docs/admin/insights.md [hunk 14]
```diff
@@ -1,35 +1,57 @@
...
+
+If no project is marked as a key project, both metrics are zero for every team, the `Community health score` loses the points they carry, and no team qualifies for the [monthly locale health report](#monthly-locale-health-report).
```
Comment (mathjazz) [inline documentation/docs/admin/insights.md]:
```suggestion
If no projects are marked as key projects, both metrics are zero for every team. Consequently, the Community health score excludes the points associated with these metrics, and no team qualifies for the [monthly locale health report](#monthly-locale-health-report).
```

File: pontoon/base/static/css/style.css
Comment (mathjazz) [reply pontoon/base/static/css/style.css]: Sorry, missed your response earlier. Could we just use `.message`?

File: pontoon/settings/base.py [hunk 15]
```diff
@@ -1235,6 +1235,12 @@ def account_username(user):
 # snapshots are collected via collect_chs_snapshots().
 MONTHLY_CHS_SNAPSHOTS_DAY = int(os.environ.get("MONTHLY_CHS_SNAPSHOTS_DAY", 1))
 
+# Minimum absolute month-over-month Community Health Score percentage change for
+# a locale to be included in the monthly locale health report notification.
```
Comment (mathjazz) [inline pontoon/settings/base.py]: It's also email...

```suggestion
# a locale to be included in the monthly locale health report.
```

File: pontoon/messaging/management/commands/send_monthly_health_report_notifications_and_emails.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "--force",
+            action="store_true",
+            help="Force compilation and send regardless of the current date.",
+        )
+
+    def handle(self, *args, **options):
+        """
+        Sends the monthly Community Health Report to all admins as a notification.
```
Comment (mathjazz) [inline pontoon/messaging/management/commands/send_monthly_health_report_notifications_and_emails.py]:
```suggestion
        Sends the monthly locale health report as a notification and email.
```

File: documentation/docs/admin/insights.md [hunk 17]
```diff
@@ -1,35 +1,57 @@
 # Insights
 
+Pontoon provides 2 major systems for admins to monitor locale & project health. The tools
+available are the `Insights` dashboard and the `Monthly locale health report`, which both provide
```
Comment (mathjazz) [inline documentation/docs/admin/insights.md]: Code identifiers don't use use word locale (e.g. `monthly_health_report`). Should we also drop it in the copy?

File: pontoon/contributors/templates/contributors/widgets/notifications_menu.html (hunk 1 above)
Comment (mathjazz) [reply pontoon/contributors/templates/contributors/widgets/notifications_menu.html]: Could we render this content under the existing `.message` class rather than restyling `.description`? The heading, paragraph, list and table styling already exists for `.message` — we'd keep the div, since block elements can't live in a span, but `UserNotification.css` wouldn't need the new block, and `style.css` wouldn't need the merged selectors either.

One consequence to weigh: the unreviewed-suggestions notification renders through the same component, so it would also pick up .message's padding and heading styling. Probably an improvement, but worth a look before and after.

**Review by mathjazz — COMMENTED** [review]
Sent copy through LLM and left some suggestions. 

The rest looks good, great work!

**Review by mathjazz — APPROVED** [review]
Looks great, thanks!
